network driver: fetch configuration from CRD during startup

bersoare opened the pull request:
improved the crd definition by swapping maps with slices, in order to simplify the json tags
on startup, the driver now loads the configuration from a CRD, and starts the driver when a configuration is found.
removed the configuration types made obsolete by the CRD
adjusted the validation of the config to validate the CRD structs

```release-note
Configure Cilium Network Driver from CRD definition
```

bersoare committed:
network driver: change schema to use slices instead of maps

maps make it hard to change the json tag for the keys,
and using an array makes the CRD looks cleaner as
we can decouple the field name from the json tag without
having to implement a custom marshaler.

also adding some default values that were missing

Signed-off-by: Bernardo Soares <20172413+bersoare@users.noreply.github.com>

bersoare committed:
network driver: use the configuration types from the CRD

no longer use the config types we were using previously,
since now we want to get the config from the crd

Signed-off-by: Bernardo Soares <20172413+bersoare@users.noreply.github.com>

bersoare committed:
network driver: fix config validation for the CRD type

adjusting the current config validation to validate
the struct from the CRD

removed the config types the code was previously using

Signed-off-by: Bernardo Soares <20172413+bersoare@users.noreply.github.com>

bersoare committed:
network driver: fetch configuration from CRD

get and use configuration from the CRD

Signed-off-by: Bernardo Soares <20172413+bersoare@users.noreply.github.com>

bersoare committed:
network driver: fix dummy filtering logic for matching ifname

Signed-off-by: Bernardo Soares <20172413+bersoare@users.noreply.github.com>

bersoare committed:
network driver: update readme

Signed-off-by: Bernardo Soares <20172413+bersoare@users.noreply.github.com>

pippolo84 commented on pkg/networkdriver/cell.go:
```diff
@@ -33,62 +36,26 @@ type networkDriverParams struct {
-//	}
-func getNetworkDriverConfig(_ k8sClient.Clientset) (*v2alpha1.CiliumNetworkDriverConfigSpec, error) {
-	return nil, nil
+func getNetworkDriverConfigResource(cs k8sClient.Clientset, lc cell.Lifecycle, mp workqueue.MetricsProvider) resource.Resource[*v2alpha1.CiliumNetworkDriverConfig] {
```
Nit: conventionally we name this functions `<resource_name>Resource`, that is `ciliumNetworkDriverConfigResource` in this case

pippolo84 commented on pkg/networkdriver/driver.go:
```diff
@@ -168,9 +169,40 @@ func (driver *Driver) publish(ctx context.Context) error {
+
+	// how long we wait for the crd config to be found on an event
+	crdTimeout := 1 * time.Second
+	crdPollInterval := 100 * time.Millisecond
```
I'm afraid that a timeout of just 1 sec might make it difficult for the driver to receive the intended config at startup.

Aside from the interval tuning, I feel that timing out this operation is not the best option. Right now, we want to rely on a CRD to have a richer configuration, possibly different for each manager type that will be enabled. Therefore, what we have to reject is, once the driver has been enabled, any run-time change to the config. Regarding the first version of the config, we should be able to accept it at any time, for increased flexibility.

I think that handling the resource events to:

1) wait for a resource sync
2) ingest the first version of CiliumNetworkDriverConfig where the driver is enabled (such version might be already available when the sync arrives, as an example in case of a restart)
3) reject further changes to CiliumNetworkDriverConfig (maybe log a warning to make the user aware of the current limitation)

is a better option.

bersoare replied:
what if there is no config in the cluster? how can we determine the difference between a configuration absent and CRD sync in terms of timing out?

pippolo84 commented on pkg/networkdriver/dummy/dummy.go:
```diff
@@ -99,10 +99,14 @@ func (d DummyDevice) Match(filter v2alpha1.CiliumNetworkDriverDeviceFilter) bool
+			}
 		}
+
+		return false
```
Nice catch, I wasn't thinking about having multiple filters here 🤦

pippolo84 commented:
Food for thought: I'm wondering if we could leverage the [Dynamic ConfigMap](https://github.com/cilium/cilium/issues/34101) and [Dynamic Lifecycle](https://github.com/cilium/cilium/pull/34936) features for this case.

bersoare committed:
network driver: refactor startup and configuration reading

agent now waits for a config at any point in time and starts
the network driver if one ever arrives.

also encapsulated the containerd and dra initialization logic
to make the start sequence more readable. no functional changes
were made

updates to configuration aren't handled yet.

Signed-off-by: Bernardo Soares <20172413+bersoare@users.noreply.github.com>

bersoare commented:
/test

bersoare commented on pkg/networkdriver/driver.go:
```diff
@@ -168,18 +170,79 @@ func (driver *Driver) publish(ctx context.Context) error {
+					driver.config = &newConfig.Spec
+				}
+
+				return nil
```
that's right - i did it intentionally and created a separate issue to handle the config update process better. otherwise, i think adding those changes in here would make this pr bigger than it needs to be.

pippolo84 commented on pkg/networkdriver/driver.go:
```diff
@@ -168,120 +172,219 @@ func (driver *Driver) publish(ctx context.Context) error {
+		for {
+			select {
+			case <-ctx.Done():
+				driver.logger.InfoContext(
```
This log is not needed, hive will log start and stop for each module for us.

pippolo84 reviewed (changes requested):
Thanks for the further iteration. I left some comments inline, but I'm afraid we had a misunderstanding on the point of my feedback [here](https://github.com/cilium/cilium/pull/43265#discussion_r2634808682).

I see you already added the handling of configuration updates. That is pretty complicated and I agree to leave it for a separate PR. I think we should strive to simplify the code more as right now is pretty hard to follow (see my last comments).

My original feedback was just to fix the code that logs a further configuration attempt, in order to warn the user that, at the moment, we are not changing the config at runtime. So what I'm suggesting is to keep this last version for a subsequent PR, maybe improving it a bit for better readability and just add the bare minimum to fix the issue pointed out in https://github.com/cilium/cilium/pull/43265#discussion_r2634808682.

I think we can solve with this:

```go
// watchConfig blocks until the first configuration is found (from the CRD).
func (driver *Driver) watchConfig(ctx context.Context) <-chan v2alpha1.CiliumNetworkDriverConfigSpec {
	ch := make(chan v2alpha1.CiliumNetworkDriverConfigSpec)

	go func() {
		defer close(ch)

		var (
			synced   bool
			received bool
		)
		for ev := range driver.configCRD.Events(ctx) {
			ev.Done(nil)

			if ev.Kind == resource.Delete {
				continue
			}

			if ev.Kind == resource.Sync {
				synced = true
				continue
			}

			// wait for sync before reading config updates
			if !synced {
				continue
			}

			// discard updates if we already received a config
			if received {
				driver.logger.InfoContext(
					ctx, "config received, but we already have one",
				)
				continue
			}
			received = true

			ch <- ev.Object.Spec
		}
	}()

	return ch
}
```

And from `Start`:

```go
		cfg := <-driver.watchConfig(ctx)
		driver.config = &cfg

		if driver.config == nil {
		  ...
```

I think this is enough for this PR to land on the feature branch and also unblock https://github.com/cilium/cilium/pull/43447 👍

bersoare committed:
network driver: log CRD update attempts

continue reading from the channel
even after the first configuration is received,
and log any further update attempts

Signed-off-by: Bernardo Soares <20172413+bersoare@users.noreply.github.com>

bersoare committed:
network driver: avoid panic due to missing configuration

check for an empty config before
accessing the config definitions (and avoid a nil
pointer access)

Signed-off-by: Bernardo Soares <20172413+bersoare@users.noreply.github.com>

bersoare commented:
/ci-eks

bersoare commented:
/ci-integration

bersoare commented:
/test

pippolo84 committed:
network-driver: Handle config after restart

In case of an agent restart without a change to the
CiliumNetworkDriverConfig we are going to receive a Sync after an Upsert
for the resource. In that case, we should read the last available config
version and enable the driver accordingly.  Conversely, if a new version
of the CiliumNetworkDriverConfig is upserted after the initial sync, we
should handle that version.

The commit updates the loop reading the events to correctly handle both
scenarios (upsert + sync and sync + upsert) while discarding further
attempts of updating the config.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 reviewed (approved):
💯 

I've added a commit to correctly handle the case when we receive a sync after an upsert for the driver config (might happen after a restart). PTAL @bersoare
